engineapi/proxy.go

```diff
@@ -19,77 +23,330 @@ import (
+		clock:        &clock.RealClock{},
+		lock:         &sync.RWMutex{},
+		stopCh:       stopCh,
+	}
+}
+
+func (m *EngineClientProxyManager) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	m.logger.Info("Start Longhorn proxy manager")
+	defer m.logger.Info("Shutting down Longhorn proxy manager")
+
+	go wait.Until(m.resync, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func getLoggerForProxyManager(logger logrus.FieldLogger, imName string) *logrus.Entry {
+	return logger.WithField("instanceManager", imName)
+}
+
+func (m *EngineClientProxyManager) resync() {
+	for {
+		for imName, handle := range m.handler.Handle {
+			log := getLoggerForProxyManager(m.logger, imName)
+
+			im, err := m.ds.GetInstanceManagerRO(handle.instanceManagerName)
+			if err != nil {
+				log.WithError(err).Warnf("Failed to get instance manager for proxy")
+				continue
+			}
+
+			for clientIndex, client := range handle.Clients {
+				if !client.ShouldResync(m.clock) {
+					continue
+				}
+
+				cLog := client.GetLogger()
+
+				if client.ShouldDelete() {
+					cLog.Debugf("Deleting proxy client")
+					client.SetNextGC(m.clock)
+
+					newIndex, err := m.getNewClientIndex(handle.Clients)
+					if err != nil {
+						cLog.WithError(err).Debug("Cannot replace proxy client")
+					} else {
+						cLog.Debug("Replacing with new proxy client")
+						m.handler.NewHandleClient(im, newIndex)
+					}
+
+					client.SetNextResync(m.clock, m.resyncPeriod)
+					break
+				}
+
+				m.gc(imName, clientIndex)
+
+				if client.ShouldCreate() {
+					cLog.Debug("Creating proxy client")
+
+					isInstanceManagerRunning := im.Status.CurrentState == longhorn.InstanceManagerStateRunning
+					if !isInstanceManagerRunning {
+						cLog.Warnf("Cannot create new gRPC client while instance manager in %v state", im.Status.CurrentState)
+					}
+
+					hasIP := im.Status.IP != ""
+					if !hasIP {
+						cLog.Warn("Cannot create new gRPC client while instance manager status IP is empty")
+					}
+
+					if err := client.Start(clientIndex, im, m.logger, m.ds, m.stopCh); err != nil {
+						cLog.Warn("Failed to start proxy client")
+					}
+				}
+
+				client.SetNextResync(m.clock, m.resyncPeriod)
+			}
+
+			exceedMaxConnection := true
+			for _, client := range handle.Clients {
+				if client.GetInUse() >= MaxProxyClientConcurrentInUse {
+					continue
+				}
+
+				exceedMaxConnection = false
+				break
+			}
+			if len(handle.Clients) != 0 && exceedMaxConnection {
+				log.Debugf("Exceed proxy client max concurrent use limit:%v", MaxProxyClientConcurrentInUse)
+
+				newIndex, err := m.getNewClientIndex(handle.Clients)
+				if err != nil {
+					log.WithError(err).Debugf("Cannot add more proxy client")
+				} else {
+					log.Debugf("Add proxy client-%v to handle concurrent use", newIndex)
+					m.handler.NewHandleClient(im, newIndex)
+				}
+			}
+		}
+	}
+}
+
+func (m *EngineClientProxyManager) getNewClientIndex(clients map[int]EngineClientProxy) (int, error) {
```

Comment:
This looks weird... Why not directly ask the handler to launch a new client if the limit is not exceeded?

Reply:
I want to ensure the client doesn't get overwritten when waiting for GC. So the new client will always use a new index while the old waits for GC.

Reply:
Use UUID instead.